pushy/src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -826,46 +802,34 @@ protected AuthenticationTokenSupplier getAuthenticationTokenSupplierForTopic(fin
-
-            if (promise != null) {
-                responsePromise = promise;
-            } else {
-                responsePromise = new DefaultPromise<>(connectionReadyPromise.channel().eventLoop());
-            }
+            final Promise<PushNotificationResponse<ApnsPushNotification>> responsePromise =
+                    new DefaultPromise<>(channel.eventLoop());
 
-            connectionReadyPromise.channel().eventLoop().submit(new Runnable() {
+            channel.eventLoop().submit(new Runnable() {
 
                 @Override
                 public void run() {
-                    if (ApnsClient.this.responsePromises.containsKey(notification)) {
-                        responsePromise.setFailure(new IllegalStateException(
-                                "The given notification has already been sent and not yet resolved."));
-                    } else {
-                        // We want to do this inside the channel's event loop so we can be sure that only one thread is
-                        // modifying responsePromises.
-                        ApnsClient.this.responsePromises.put(notification, responsePromise);
-                    }
-                }
-            });
+                    final ApnsClientHandler handler = channel.pipeline().get(ApnsClientHandler.class);
 
-            connectionReadyPromise.channel().writeAndFlush(notification).addListener(new GenericFutureListener<ChannelFuture>() {
-
-                @Override
-                public void operationComplete(final ChannelFuture future) throws Exception {
-                    if (future.isSuccess()) {
-                        ApnsClient.this.metricsListener.handleNotificationSent(ApnsClient.this, notificationId);
-                    } else {
-                        log.debug("Failed to write push notification: {}", notification, future.cause());
+                    try {
+                        handler.setResponsePromiseForNotification(notification, responsePromise);
+                        channel.writeAndFlush(notification).addListener(new GenericFutureListener<ChannelFuture>() {
 
-                        // This will always be called from inside the channel's event loop, so we don't have to worry
-                        // about synchronization.
-                        ApnsClient.this.responsePromises.remove(notification);
-                        responsePromise.tryFailure(future.cause());
+                            @Override
+                            public void operationComplete(final ChannelFuture future) throws Exception {
+                                if (future.isSuccess()) {
+                                    ApnsClient.this.metricsListener.handleNotificationSent(ApnsClient.this, notificationId);
+                                }
+                            }
+                        });
+                    } catch (final IllegalStateException e) {
```

Comment:
I'm debating making this a more purpose-specific exception. On one hand, it seems bogus to have a ton of super narrowly-scoped exceptions (maybe they should have common base class?), but on the other, we might get `IllegalStateExceptions` for other reasons.